src/unix/index_unix.ml

```diff
@@ -254,6 +254,34 @@ module IO : Index.IO = struct
       let _ = Unix.fstat t.raw.fd in
       true
     with Unix.Unix_error (Unix.EBADF, _, _) -> false
+
+  type fd = Unix.file_descr
+
+  let _lock op f =
+    mkdir (Filename.dirname f);
+    let fd = Unix.openfile f [ Unix.O_CREAT; Unix.O_RDWR ] 0o600
+    and pid = string_of_int (Unix.getpid ()) in
+    let pid_len = String.length pid in
+    try
+      Unix.lockf fd op 0;
+      if Unix.single_write_substring fd pid 0 pid_len <> pid_len then (
+        Unix.close fd;
+        failwith "Unable to write PID to lock file" )
+      else Some fd
+    with
+    | Unix.Unix_error (Unix.EAGAIN, _, _) ->
+        Unix.close fd;
+        None
+    | e ->
+        Unix.close fd;
+        raise e
+
+  let lock path =
+    match _lock Unix.F_LOCK path with
+    | Some fd -> fd
+    | None -> failwith "lockf F_LOCK didn't lock"
```

Comment:
maybe use a more informative error message, including the name of the lock file?